File: design_patterns/src/prototype/go/c1.go
package main

import "log"

/*
原型模式
 */

type Prototype interface {
  setX()
  ssetY()
  Clone() *Object
}

type Object struct {
  x     uint8
  y     uint8
}

func (o *Object) setX() {
  panic("implement me")
}

func (o *Object) ssetY() {
  panic("implement me")
}

func (o *Object) Clone() *Object {
  oc := *o
  return &oc      // 浅拷贝
}

func NewObject() Prototype {    // todo: 相当于在 rust里面 impl trait for struct
  return &Object{
    x: 100,
    y: 200,
  }
}

func main() {
  origin := NewObject()
  log.Printf("origin --------------- %+v, %+v", &origin, origin)
  objc := origin.Clone()
  log.Printf("objc --------------- %+v, %+v", &objc, objc)
}

File: design_patterns/src/prototype/c1.rs
/*
原型模式
 */

trait Prototype: Clone {
  fn set_x(&mut self, _: usize);
  fn set_y(&mut self, _: usize);
}

#[derive(Debug, Clone)]
struct Object {
  x:  usize,
  y:  usize,
}

impl Object {
  fn new() -> Object {
    Object {
      x:    100,
      y:    200,
    }
  }
}

impl Prototype for Object {
  fn set_x(&mut self, x: usize) {
    self.x = x
  }

  fn set_y(&mut self, y: usize) {
    self.y = y
  }
}

fn main() {
  let origin = Object::new();
  // let origin_addr = origin as *const i32;
  // let origin_addr = &origin;
  println!("origin ptr addr: {:p}", &origin);

  let mut obj = origin.clone();
  obj.set_x(39);
  println!("obj ptr addr: {:p}", &obj);

  println!("origin = {:?}", origin);
  println!("obj = {:?}", obj);
}





